synapse/handlers/sync.py

```diff
@@ -494,10 +495,105 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
+        state events to 'state' if needed to describe the heros.
+
+        Args:
+            room_id(str):
+            sync_config(synapse.handlers.sync.SyncConfig):
+            batch(synapse.handlers.sync.TimelineBatch): The timeline batch for
+                the room that will be sent to the user.
+            state(dict): dict of (type, state_key) -> Event as returned by
+                compute_state_delta
+            now_token(str): Token of the end of the current batch.
+            full_state(bool): Whether to force returning the full state.
+
+        Returns:
+             A deferred dict describing the room summary
+        """
+
+        # FIXME: this promulgates https://github.com/matrix-org/synapse/issues/3305
+        last_events, _ = yield self.store.get_recent_events_for_room(
+            room_id, end_token=now_token.room_key, limit=1,
+        )
+
+        if not last_events:
+            defer.returnValue(None)
+            return
+
+        last_event = last_events[-1]
+        state_ids = yield self.store.get_state_ids_for_event(
+            last_event.event_id, [
+                (EventTypes.Member, None),
+                (EventTypes.Name, ''),
+                (EventTypes.CanonicalAlias, ''),
+            ]
+        )
+
+        member_ids = {
+            state_key: event_id
+            for (t, state_key), event_id in state_ids.iteritems()
+            if t == EventTypes.Member
+        }
+        name_id = state_ids.get((EventTypes.Name, ''))
+        canonical_alias_id = state_ids.get((EventTypes.CanonicalAlias, ''))
+
+        summary = {}
+
+        # ideally we could do this, but the actual events haven't been loaded...
+        # summary['m.joined_member_count'] = len([
+        #     s for s in member_ids if s.content.get("membership") == 'join'
+        # ])
+
+        # FIXME: this includes left/banned/invited users
+        summary['m.joined_member_count'] = len(member_ids)
+
+        if not name_id and not canonical_alias_id:
+            # FIXME: this includes left/banned/invited users
+            # FIXME: if joined_member_count is 0, return left users instead
+            # FIXME: order by stream ordering, not alphabetic
+            # FIXME: exclude the current logged in user
+            summary['m.heros'] = sorted(member_ids.keys())[0:5]
+
+            # ensure we send membership events for heros if needed
+            cache_key = (sync_config.user.to_string(), sync_config.device_id)
+            cache = self.lazy_loaded_members_cache.get(cache_key)
+
+            existing_members = {
+                user_id: True for (typ, user_id) in state.keys()
+                if typ == EventTypes.Member
+            }
+
+            for ev in batch.events:
+                if ev.type == EventTypes.Member:
+                    existing_members[ev.state_key] = True
+
+            missing_hero_event_ids = [
+                member_ids[hero_id]
+                for hero_id in summary['heros']
```

Comment:
Should be `m.heros`? I get a `exceptions.KeyError: 'heros'` when running this locally.

Reply:
Fixed it, as am 99% certain that that is what it should be, and it unblocked me locally from syncing.